source/develop/release-management/features/storage/incremental-backup.html.md

```diff
@@ -28,138 +25,246 @@ allow incremental backup to start.
+   is ready.
+1. Backup application starts a transfer with the relevant disks
+1. Backup application obtains the changed blocks list from ovirt-imageio
+   via HTTPS.
+1. ovirt-imageio obtains the changed blocks using qemu-img map and the
+   external image.
+1. Backup application downloads the changed blocks (guest blocks) from
+   ovirt-imageio via HTTPS.
+1. ovirt-imageio reads the blocks from /dev/nbdX.
+
+
+### Finalizing backup
+
+When backup succeeds, backup application completed to copy all data.
+
+1. Delete previous backup checkpoint using libvirt API
+2. Delete scratch disks
+3. If the vm is not running, tear down disks
+
+
+### Cancelling backup
+
+When backup fail or aborted, backup application did completed to copy
+all data.
+
+1. If backup was started in libvirt, delete current backup checkpoint
+   using libvirt API
+2. Delete scratch disks
+3. If the vm is not running, tear down disks
+
+
+### Engine database changes
+
+- New ```vm_backups``` table
+  - fields: id, vm_id, incremental, creation_date, status
+  - status: initializing, ready, transferring, finallizing
+- ```images``` table
+  new field: ```backup_id``` (foreign key ```bakcups:id```)
+  new field: ```backup_url``` (nbd url used in image transfer ticket)
+
+
+### REST
+
+
+#### Starting backup
+
+Start incremental backup since backup id "previous-backup-uuid".
+
+Request:
+```
+POST /vms/vm-uuid/backups
+
+<backup>
+    <incremental>previous-backup-uuid</incremental>
+    <disks>
+        <disk id="disk-uuid" />
+        ...
+    </disks>
+</backup>
+```
+
+Response:
+```
+<backup id="backup-uuid">
+    <incremental>previous-backup-uuid</incremental>
+    <disks>
+        <disk id="disk-uuid" />
+        ...
+        ...
+    </disks>
+    <status>initiailizing</status>
+    <creation_date>...
+</backup>
+```
+
+#### Getting specific backup
+
+When backup status is "ready", you can start downloading the disks.
+
+Request:
+```
+GET /vms/vm-uuid/backups/backup-uuid
+```
+
+Response:
+```
+<backup id="backup-uuid">
+    <incremental>previous-backup-uuid</incremental>
+    <disks>
+        <disk id="disk-uuid">
+            <image_id>image-uuid</image_id>
+        </disk>
+        ...
+    </disks>
+    <status>ready</status>
+    <creation_date>...
+</backup>
+```
+
+#### Finalizing backup
+
+```
+POST /vms/vm-uuid/backups/backup-uuid/finalize
+
+<action></action>
+```
+
+#### Canceling backup
+
+```
+POST /vms/vm-uuid/backups/backup-uuid/cancel
+
+<action></action>
+```
+
+
+### Incremental backup ticket example
+
+
+For running VM, qemu will serve the disk using NBD:
+```
+{
+    "url": "nbd:localhost:1234:exportname=/sda"
+}
+```
+
+For stopped VM, we will run one qemu-nbd instance per disk, using unix
+socket:
+```
+{
+    "url": "nbd:unix:/run/vdsm/nbd/xxxyyy-nbd.sock:exportname=/sda"
+}
+```
+
+### Open Issues
+
+- What should we do when VM is stopped? Cancel backup flow?
+
+- How to handle entities locks upon backup? 
+
+- How to handle failure on finalize?
```

Comment:
We should add that if a user wants to create a backup checkpoints on RAW format, we create a qcow on top of that disk automatically (a qcow on top that should not be removed/managed any other way than the CBT setting). 
This is because we abstract the qcow/raw formats from the user and he need to be able to control enablement of CBT per disk.

Reply:
I'm not sure how this can work. To add new layer on a disk, we need to add a new snapshot. If you remove all snapshots, you will loose the ability to do incremental backups.

So maybe we need to add specify something like:

    [ ] Enable incremental backup 

When adding a disk. If this disk is enabled for incremental backup,
we will allow creating preallocated disks.

Anyway we will not have a way to enable incremental backup for direct LUN or cinderlib based disk unless we
add the option of using qcow2 format on these disks.

I think this is something that we will have to work after we have a working backup for qcow2 disks.

I'll add open issues section for such issues.